client/README.md

```diff
@@ -1,27 +1,58 @@
-# Longhorn API Client
+# Longhorn API client
 
-The code in this package are generated by [rancher/go-rancher](https://github.com/rancher/go-rancher).
-But in order to fix some compatibility issues, you need to use this [repo](https://github.com/niusmallnan/go-rancher), and you should checkout the `longhorn-v1` branch.
+The code in this package is generated by [rancher/go-rancher](https://github.com/rancher/go-rancher). But in order to
+fix some compatibility issues, you must use the `longhorn-v1` branch of this
+[repo](https://github.com/niusmallnan/go-rancher) instead.
 
 ## Generate code in go-rancher
 
-If you have a longhorn api URL, you can use this script:
+The version of go-rancher we use does not have Go module support. It must be cloned onto the correct location on the
+GOPATH and the Go tools must be executed with `GO111MODULE=off`. This might be unusual for developers used to working
```

Comment:
Looks like I am able to run it without `GO111MODULE=off`. The path need to be under GOPATH indeed

Reply:
Sorry @PhanLe1010, can you clarify whether you are agreeing with this addition, or if changes need to me made? Thanks.

Reply:
Ah, I misread your comment. I don't remember that working for me. I will check again. What version is your Go binary?

- [ ] Verify whether or not we need GO111MODULE=off.

Reply:
These are my results:

```
eweber@laptop:~/go/src/github.com/niusmallnan/go-rancher> go version
go version go1.20.2 linux/amd64

eweber@laptop:~/go/src/github.com/niusmallnan/go-rancher> go get
go: go.mod file not found in current directory or any parent directory.
        'go get' is no longer supported outside a module.
        To build and install a command, use 'go install' with a version,
        like 'go install example.com/cmd@latest'
        For more information, see https://golang.org/doc/go-get-install-deprecation
        or run 'go help get' or 'go help install'.

eweber@laptop:~/go/src/github.com/niusmallnan/go-rancher> ./scripts/generate-longhorn-schemas.sh http://164.92.95.242:30186
++ dirname ./scripts/generate-longhorn-schemas.sh
+ cd ./scripts/../generator
+ URL_BASE=http://localhost:8080
+ '[' 1 -gt 1 ']'
+ '[' http://164.92.95.242:30186 '!=' '' ']'
+ URL_BASE=http://164.92.95.242:30186
+ source ../scripts/common_functions
++ set -e
+ gen http://164.92.95.242:30186/v1 longhorn rename
+ BASE=http://164.92.95.242:30186/v1
+ curl -s 'http://164.92.95.242:30186/v1/schemas?_role=service'
+ jq .
+ echo Saved schemas.json
Saved schemas.json
+ echo -n Generating go code...
Generating go code...+ rm -rf '../output/generated_*'
+ go run generator.go
generator.go:13:2: no required module provides package github.com/rancher/go-rancher/client: go.mod file not found in current directory or any parent directory; see 'go help modules'
```

Can't get dependencies or generate code due to modules related errors.

With GO111MODULE=off:

```
eweber@laptop:~/go/src/github.com/niusmallnan/go-rancher> GO111MODULE=off go get

eweber@laptop:~/go/src/github.com/niusmallnan/go-rancher> GO111MODULE=off ./scripts/generate-longhorn-schemas.sh http://164.92.95.242:30186
++ dirname ./scripts/generate-longhorn-schemas.sh
+ cd ./scripts/../generator
+ URL_BASE=http://localhost:8080
+ '[' 1 -gt 1 ']'
+ '[' http://164.92.95.242:30186 '!=' '' ']'
+ URL_BASE=http://164.92.95.242:30186
+ source ../scripts/common_functions
++ set -e
+ gen http://164.92.95.242:30186/v1 longhorn rename
+ BASE=http://164.92.95.242:30186/v1
+ curl -s 'http://164.92.95.242:30186/v1/schemas?_role=service'
+ jq .
+ echo Saved schemas.json
Saved schemas.json
+ echo -n Generating go code...
Generating go code...+ rm -rf '../output/generated_*'
+ go run generator.go
+ echo ' Done'
 Done
+ gofmt -w ../output/generated_activate_input.go ../output/generated_api_version.go ../output/generated_attach_input.go ../output/generated_attachment.go ../output/generated_backing_image_cleanup_input.go ../output/generated_backing_image_disk_file_status.go ../output/generated_backing_image.go ../output/generated_backup.go ../output/generated_backup_input.go ../output/generated_backup_list_output.go ../output/generated_backup_status.go ../output/generated_backup_target.go ../output/generated_backup_volume.go ../output/generated_client.go ../output/generated_clone_status.go ../output/generated_controller.go ../output/generated_detach_input.go ../output/generated_disk_condition.go ../output/generated_disk_info.go ../output/generated_disk_update.go ../output/generated_disk_update_input.go ../output/generated_empty.go ../output/generated_engine_image.go ../output/generated_engine_upgrade_input.go ../output/generated_error.go ../output/generated_event.go ../output/generated_expand_input.go ../output/generated_instance_manager.go ../output/generated_instance_process.go ../output/generated_kubernetes_status.go ../output/generated_longhorn_condition.go ../output/generated_node_condition.go ../output/generated_node.go ../output/generated_orphan.go ../output/generated_purge_status.go ../output/generated_pvccreate_input.go ../output/generated_pvcreate_input.go ../output/generated_rebuild_status.go ../output/generated_recurring_job.go ../output/generated_replica.go ../output/generated_replica_remove_input.go ../output/generated_restore_status.go ../output/generated_salvage_input.go ../output/generated_setting_definition.go ../output/generated_setting.go ../output/generated_snapshot_cr.go ../output/generated_snapshot_crinput.go ../output/generated_snapshot_crlist_output.go ../output/generated_snapshot.go ../output/generated_snapshot_input.go ../output/generated_snapshot_list_output.go ../output/generated_support_bundle.go ../output/generated_support_bundle_initate_input.go ../output/generated_system_backup.go ../output/generated_system_restore.go ../output/generated_tag.go ../output/generated_update_access_mode_input.go ../output/generated_update_backup_compression_input.go ../output/generated_update_data_locality_input.go ../output/generated_update_offline_replica_rebuilding_input.go ../output/generated_update_replica_auto_balance_input.go ../output/generated_update_replica_count_input.go ../output/generated_update_replica_disk_soft_anti_affinity_input.go ../output/generated_update_replica_soft_anti_affinity_input.go ../output/generated_update_replica_zone_soft_anti_affinity_input.go ../output/generated_update_snapshot_data_integrity_input.go ../output/generated_update_unmap_mark_snap_chain_removed_input.go ../output/generated_volume_attachment.go ../output/generated_volume_condition.go ../output/generated_volume.go ../output/generated_volume_recurring_job.go ../output/generated_volume_recurring_job_input.go ../output/generated_workload_status.go
+ echo Formatted code
Formatted code
+ '[' -n rename ']'
+ rm -rf ../longhorn/generated_activate_input.go ../longhorn/generated_api_version.go ../longhorn/generated_attach_input.go ../longhorn/generated_attachment.go ../longhorn/generated_backing_image_cleanup_input.go ../longhorn/generated_backing_image_disk_file_status.go ../longhorn/generated_backing_image.go ../longhorn/generated_backup.go ../longhorn/generated_backup_input.go ../longhorn/generated_backup_list_output.go ../longhorn/generated_backup_status.go ../longhorn/generated_backup_target.go ../longhorn/generated_backup_volume.go ../longhorn/generated_client.go ../longhorn/generated_clone_status.go ../longhorn/generated_controller.go ../longhorn/generated_detach_input.go ../longhorn/generated_disk_condition.go ../longhorn/generated_disk_info.go ../longhorn/generated_disk_update.go ../longhorn/generated_disk_update_input.go ../longhorn/generated_empty.go ../longhorn/generated_engine_image.go ../longhorn/generated_engine_upgrade_input.go ../longhorn/generated_error.go ../longhorn/generated_event.go ../longhorn/generated_expand_input.go ../longhorn/generated_instance_manager.go ../longhorn/generated_instance_process.go ../longhorn/generated_kubernetes_status.go ../longhorn/generated_longhorn_condition.go ../longhorn/generated_node_condition.go ../longhorn/generated_node.go ../longhorn/generated_orphan.go ../longhorn/generated_purge_status.go ../longhorn/generated_pvccreate_input.go ../longhorn/generated_pvcreate_input.go ../longhorn/generated_rebuild_status.go ../longhorn/generated_recurring_job.go ../longhorn/generated_replica.go ../longhorn/generated_replica_remove_input.go ../longhorn/generated_restore_status.go ../longhorn/generated_salvage_input.go ../longhorn/generated_setting_definition.go ../longhorn/generated_setting.go ../longhorn/generated_snapshot_cr.go ../longhorn/generated_snapshot_crinput.go ../longhorn/generated_snapshot_crlist_output.go ../longhorn/generated_snapshot.go ../longhorn/generated_snapshot_input.go ../longhorn/generated_snapshot_list_output.go ../longhorn/generated_support_bundle.go ../longhorn/generated_support_bundle_initate_input.go ../longhorn/generated_system_backup.go ../longhorn/generated_system_restore.go ../longhorn/generated_tag.go ../longhorn/generated_update_access_mode_input.go ../longhorn/generated_update_backup_compression_input.go ../longhorn/generated_update_data_locality_input.go ../longhorn/generated_update_offline_replica_rebuilding_input.go ../longhorn/generated_update_replica_auto_balance_input.go ../longhorn/generated_update_replica_count_input.go ../longhorn/generated_update_replica_disk_soft_anti_affinity_input.go ../longhorn/generated_update_replica_soft_anti_affinity_input.go ../longhorn/generated_update_replica_zone_soft_anti_affinity_input.go ../longhorn/generated_update_snapshot_data_integrity_input.go ../longhorn/generated_update_unmap_mark_snap_chain_removed_input.go ../longhorn/generated_volume_attachment.go ../longhorn/generated_volume_condition.go ../longhorn/generated_volume.go ../longhorn/generated_volume_recurring_job.go ../longhorn/generated_volume_recurring_job_input.go ../longhorn/generated_workload_status.go
+ mv ../output/generated_activate_input.go ../output/generated_api_version.go ../output/generated_attach_input.go ../output/generated_attachment.go ../output/generated_backing_image_cleanup_input.go ../output/generated_backing_image_disk_file_status.go ../output/generated_backing_image.go ../output/generated_backup.go ../output/generated_backup_input.go ../output/generated_backup_list_output.go ../output/generated_backup_status.go ../output/generated_backup_target.go ../output/generated_backup_volume.go ../output/generated_client.go ../output/generated_clone_status.go ../output/generated_controller.go ../output/generated_detach_input.go ../output/generated_disk_condition.go ../output/generated_disk_info.go ../output/generated_disk_update.go ../output/generated_disk_update_input.go ../output/generated_empty.go ../output/generated_engine_image.go ../output/generated_engine_upgrade_input.go ../output/generated_error.go ../output/generated_event.go ../output/generated_expand_input.go ../output/generated_instance_manager.go ../output/generated_instance_process.go ../output/generated_kubernetes_status.go ../output/generated_longhorn_condition.go ../output/generated_node_condition.go ../output/generated_node.go ../output/generated_orphan.go ../output/generated_purge_status.go ../output/generated_pvccreate_input.go ../output/generated_pvcreate_input.go ../output/generated_rebuild_status.go ../output/generated_recurring_job.go ../output/generated_replica.go ../output/generated_replica_remove_input.go ../output/generated_restore_status.go ../output/generated_salvage_input.go ../output/generated_setting_definition.go ../output/generated_setting.go ../output/generated_snapshot_cr.go ../output/generated_snapshot_crinput.go ../output/generated_snapshot_crlist_output.go ../output/generated_snapshot.go ../output/generated_snapshot_input.go ../output/generated_snapshot_list_output.go ../output/generated_support_bundle.go ../output/generated_support_bundle_initate_input.go ../output/generated_system_backup.go ../output/generated_system_restore.go ../output/generated_tag.go ../output/generated_update_access_mode_input.go ../output/generated_update_backup_compression_input.go ../output/generated_update_data_locality_input.go ../output/generated_update_offline_replica_rebuilding_input.go ../output/generated_update_replica_auto_balance_input.go ../output/generated_update_replica_count_input.go ../output/generated_update_replica_disk_soft_anti_affinity_input.go ../output/generated_update_replica_soft_anti_affinity_input.go ../output/generated_update_replica_zone_soft_anti_affinity_input.go ../output/generated_update_snapshot_data_integrity_input.go ../output/generated_update_unmap_mark_snap_chain_removed_input.go ../output/generated_volume_attachment.go ../output/generated_volume_condition.go ../output/generated_volume.go ../output/generated_volume_recurring_job.go ../output/generated_volume_recurring_job_input.go ../output/generated_workload_status.go ../longhorn
+ sed -i -e 's/package client/package longhorn/g' ../longhorn/client.go ../longhorn/common.go ../longhorn/generated_activate_input.go ../longhorn/generated_api_version.go ../longhorn/generated_attach_input.go ../longhorn/generated_attachment.go ../longhorn/generated_backing_image_cleanup_input.go ../longhorn/generated_backing_image_disk_file_status.go ../longhorn/generated_backing_image.go ../longhorn/generated_backup.go ../longhorn/generated_backup_input.go ../longhorn/generated_backup_list_output.go ../longhorn/generated_backup_status.go ../longhorn/generated_backup_target.go ../longhorn/generated_backup_volume.go ../longhorn/generated_client.go ../longhorn/generated_clone_status.go ../longhorn/generated_controller.go ../longhorn/generated_detach_input.go ../longhorn/generated_disk_condition.go ../longhorn/generated_disk_info.go ../longhorn/generated_disk_update.go ../longhorn/generated_disk_update_input.go ../longhorn/generated_empty.go ../longhorn/generated_engine_image.go ../longhorn/generated_engine_upgrade_input.go ../longhorn/generated_error.go ../longhorn/generated_event.go ../longhorn/generated_expand_input.go ../longhorn/generated_instance_manager.go ../longhorn/generated_instance_process.go ../longhorn/generated_kubernetes_status.go ../longhorn/generated_longhorn_condition.go ../longhorn/generated_node_condition.go ../longhorn/generated_node.go ../longhorn/generated_orphan.go ../longhorn/generated_purge_status.go ../longhorn/generated_pvccreate_input.go ../longhorn/generated_pvcreate_input.go ../longhorn/generated_rebuild_status.go ../longhorn/generated_recurring_job.go ../longhorn/generated_replica.go ../longhorn/generated_replica_remove_input.go ../longhorn/generated_restore_status.go ../longhorn/generated_salvage_input.go ../longhorn/generated_setting_definition.go ../longhorn/generated_setting.go ../longhorn/generated_snapshot_cr.go ../longhorn/generated_snapshot_crinput.go ../longhorn/generated_snapshot_crlist_output.go ../longhorn/generated_snapshot.go ../longhorn/generated_snapshot_input.go ../longhorn/generated_snapshot_list_output.go ../longhorn/generated_support_bundle.go ../longhorn/generated_support_bundle_initate_input.go ../longhorn/generated_system_backup.go ../longhorn/generated_system_restore.go ../longhorn/generated_tag.go ../longhorn/generated_update_access_mode_input.go ../longhorn/generated_update_backup_compression_input.go ../longhorn/generated_update_data_locality_input.go ../longhorn/generated_update_offline_replica_rebuilding_input.go ../longhorn/generated_update_replica_auto_balance_input.go ../longhorn/generated_update_replica_count_input.go ../longhorn/generated_update_replica_disk_soft_anti_affinity_input.go ../longhorn/generated_update_replica_soft_anti_affinity_input.go ../longhorn/generated_update_replica_zone_soft_anti_affinity_input.go ../longhorn/generated_update_snapshot_data_integrity_input.go ../longhorn/generated_update_unmap_mark_snap_chain_removed_input.go ../longhorn/generated_volume_attachment.go ../longhorn/generated_volume_condition.go ../longhorn/generated_volume.go ../longhorn/generated_volume_recurring_job.go ../longhorn/generated_volume_recurring_job_input.go ../longhorn/generated_workload_status.go ../longhorn/schemas.go ../longhorn/types.go
+ rm -rf ../output
+ rm schemas.json
+ rm -rf '../longhorn/*.go-e'
+ echo Success
Success
```

Is it possible you are running an older version of Go than me (`v1.20.2`) or have `GO111MODULE=off` permanently set in your Go environment?

Reply:
It is possible that I was running old version of Go. Thank you